clang/test/SemaSYCL/pointer-to-vla.cpp

```diff
@@ -0,0 +1,15 @@
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -triple spir64 -fsyntax-only -Wno-sycl-2017-compat -verify %s
+//
+// This test checks if compiler reports compilation error on an attempt to pass
+// a pointer to VLA as kernel argument
+
+#include "Inputs/sycl.hpp"
+
+void foo(unsigned X) {
+  using VLATy = float(*)[X];
```

Comment:
So VLAs are already handled in current code? i.e. int arr[X] will throw required diagnostic here?

I think zero sized array is used to implement variable length array in structs. Is that currently handled? 

```
struct test {
   int a;
   int arr[0];
} 
```

Reply:
Yes, VLAs are handled here https://github.com/intel/llvm/blob/76fa6b3bcade074bdedef740001c4528e1aa08a8/clang/lib/Sema/SemaSYCL.cpp#L1162 . However this diagnostic won't work for zero sized array since it is considered as constant size array as well.
I would say that our types diagnosing is not sufficient, because it mostly shows up if some variable of unsupported type is declared, so even kernel argument with unsupported type (like 128 bit float for example) is not diagnosed. However, I think it is a separate task because I think that we should implement types diagnosing in a bit other way. So I just implemented a diagnostic for pointers to VLAs because only they cause crash in front-end.

Reply:
Hmm.. yea. Do you know why checkSYCLType doesn't catch these? I see it is called in checkSYCLDeviceVarDecl. Is this not called earlier in Sema?

Reply:
Yes, `checkSYCLDeviceVarDecl` allows to check only **variable declarations**. That is it, there is no other place where `checkSYCLType` is used except one place in SemaSYCL where it checks type of accessor kernel arg.
Having said that, you probably wonder why I didn't just use `checkSYCLType` for my case. Well, I think having optimizations enabled (including DAE) we should move to diagnosing on attempt to *use* declaration of unsupported type, not on attempt to *declare* it. And I've already started some similar work in community code, so it actually gives a diagnostic if declaration of some long enough double or int type is *used* in device code. However it diagnoses not so much use cases as `checkSYCLType` because it is shared between SYCL and OpenMP, but it still shares some cases with `checkSYCLType`. In addition usage of `checkSYCLType` here results in double diagnosing because capturing of some unsupported typed declaration makes it a case when unsupported type is *used* (emits community diagnostic) and it is a kernel argument (so `checkSYCLType` will work here as well).
That is why I consider it as other task, here my job was prevent crash, not improve the whole diagnosing of unsupported types.